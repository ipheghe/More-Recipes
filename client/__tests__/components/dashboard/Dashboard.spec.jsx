import expect from 'expect';
import React from 'react';
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import render from 'react-test-renderer';
import { HashRouter as Router } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ConnectedDashboard, { PureDashboard }
  from '../../../src/components/Dashboard';
import mockItems from '../../__mocks__/mockItems';
import mockAuthCheck from '../../__mocks__/mockAuthCheck';


const middleware = [thunk];
const mockStore = configureMockStore(middleware);
const initialState = {
  auth: {
    error: 'Invalid username',
    message: '',
    userData: {},
    authenticated: true,
    categories: []
  },
  user: {
    status: '',
    error: 'Fail',
    message: '',
    userData: mockItems.user,
  },
  recipe: {
    message: '',
    error: '',
    pages: 2,
    recipeData: {},
    recipeList: mockItems.recipeArray,
    userRecipes: [],
    searchResult: []
  },
  category: {
    error: '',
    message: '',
    categoryList: [mockItems.category],
    userCategoryList: [mockItems.category]
  },
  imageUploadReducer: [{
    imageData: {},
    response: '/assets/images/pizza.jpg',
    error: '',
    isloaded: false
  }],
  favorite: {
    message: '',
    status: false,
    error: '',
    pages: 1,
    userFavorites: mockItems.recipeArray,
    userFavorite: [],
  }
};

const store = mockStore(initialState);

const props = {
  isAuthenticated: true,
  auth: () => true,
  location: {
    search: '',
    pathname: '/'
  },
};

/**
 * @param { boolean } isAuthenticated
 * @param { number } index
 *
 * @return { * } null
 */
const setup = (isAuthenticated, index) => {
  props.isAuthenticated = isAuthenticated;
  const mountedWrapper = mount(<Provider store={store}>
    <Router><ConnectedDashboard {...props} /></Router>
                               </Provider>);
  const shallowWrapper = shallow(<PureDashboard {...props} />);
  const wrapper = mount(<Provider store={store}>
    <MemoryRouter
      initialEntries={
        [
          '/dashboard/top-recipes',
          '/dashboard/my-recipes',
          '/Dashboard/add-recipe',
          '/dashboard/favorites',
          '/dashboard/manage-recipes',
          '/dashboard/',
          '/dashboard/top-recipes/*',
          '/dashboard/favorites/*',
          '/dashboard/manage-recipes/*',
          '/dashboard/search'
        ]
      }
      initialIndex={index}
      keyLength={6}
    >
      <PureDashboard {...props} />
    </MemoryRouter>
                        </Provider>);

  return {
    mountedWrapper,
    shallowWrapper,
    wrapper
  };
};

describe('<Dashboard', () => {
  beforeEach(() => {
    mockAuthCheck();
  });
  it('renders without crashing', () => {
    const { mountedWrapper } = setup(true, 0);
    expect(mountedWrapper).toBeDefined();
    expect(mountedWrapper.find('ProfileHeader').length).toBe(1);
    expect(mountedWrapper.find('UserSection').length).toBe(1);
    expect(mountedWrapper.find('UserNavMenu').length).toBe(1);
    expect(mountedWrapper.find('Route').length).toBe(15);
    expect(mountedWrapper.exists()).toBe(true);
  });

  it('should match component snapshot', () => {
    const tree = render.create(<Provider store={store}>
      <Router ><PureDashboard {...props} /></Router>
                               </Provider>);
    expect(tree).toMatchSnapshot();
  });

  it('renders Top Recipes componet if top-recipes route is called', () => {
    const { wrapper } = setup(false, 0);
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });

  it(`redirects from top recipes component to login 
     component if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 0);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it('renders My Recipes component if my-recipes route is called', () => {
    const { wrapper } = setup(false, 1);
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });


  it(`redirects from my recipes component to login
      component if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 1);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it('renders Add Recipe component if add-recipe route is called', () => {
    const { wrapper } = setup(false, 2);
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });

  it(`redirects from add recipe component to login
      component if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 2);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it('renders Favorite component if favorite route is called', () => {
    const { wrapper } = setup(false, 3);
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });

  it(`redirects from favorite component to login 
     component if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 3);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it(`renders Manage Recipes component 
     if manage-recipes route is called`, () => {
      const { wrapper } = setup(false, 4);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it(`redirects from manage recipes component to login 
     component if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 4);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it(`redirects to notFound page 
     if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 5);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it(`redirects to notFound page 
     if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 6);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it(`redirects to notFound page 
     if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 7);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it(`redirects to notFound page 
     if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 8);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });

  it('renders Search component if a non-existent route is called', () => {
    const { wrapper } = setup(false, 9);
    expect(wrapper).toBeDefined();
    expect(wrapper.exists()).toBe(true);
  });

  it(`redirects from Search component to login 
     component if a non-existent route is called`, () => {
      const { wrapper } = setup(true, 9);
      expect(wrapper).toBeDefined();
      expect(wrapper.exists()).toBe(true);
    });
});
