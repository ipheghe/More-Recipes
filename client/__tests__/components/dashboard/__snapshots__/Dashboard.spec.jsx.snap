// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard should match component snapshot 1`] = `
<div>
  <div
    className="banner-background"
  >
    <div
      className="profile-background"
    >
      <div
        className="container"
      >
        <nav
          className="navbar navbar-toggleable-md navbar-light bg-faded"
        >
          <button
            aria-controls="navbarTogglerDemo01"
            aria-expanded="false"
            aria-label="Toggle navigation"
            className="navbar-toggler"
            data-target="#navbarTogglerDemo01"
            data-toggle="collapse"
            type="button"
          >
            <span
              className="navbar-toggler-icon"
            />
          </button>
          <div
            className="collapse navbar-collapse"
            id="navbarTogglerDemo01"
          >
            <a
              className="navbar-brand invisible-button"
              href="#/dashboard/top-recipes"
              onClick={[Function]}
            >
              <h2>
                Profile Page
              </h2>
            </a>
            <ul
              className="navbar-nav mr-auto mt-2 mt-lg-0"
            >
              <li
                className="nav-item"
              >
                <button
                  className="nav-link invisible-button disabled"
                />
              </li>
            </ul>
            <form
              className="form-inline my-2 my-lg-0"
            >
              <input
                className="form-control mr-sm-2"
                formNoValidate={true}
                name="keyword"
                onChange={[Function]}
                placeholder="Search"
                required={true}
                type="text"
                value=""
              />
              <button
                className="btn btn-outline-success my-2 my-sm-0"
                id="search-recipe"
                onClick={[Function]}
              >
                Search
              </button>
            </form>
          </div>
        </nav>
        <br />
        <div
          className="row profile-landing"
        >
          <section
            className="col-md-3 profile-details"
          >
            <main>
              <div
                className="div-profile"
              >
                <h5>
                  @
                </h5>
                <p>
                  <span>
                    <i
                      aria-hidden="true"
                      className="fa fa-circle"
                      style={
                        Object {
                          "color": "green",
                        }
                      }
                    />
                  </span>
                  Online
                </p>
                <br />
                <br />
                <button
                  className="invisible-button nav-editProfile"
                  onClick={[Function]}
                  style={
                    Object {
                      "color": "#0275d8",
                      "textAlign": "center",
                    }
                  }
                >
                  Edit Profile
                </button>
                <button
                  className="invisible-button nav-changePassword"
                  onClick={[Function]}
                  style={
                    Object {
                      "color": "#0275d8",
                      "textAlign": "center",
                    }
                  }
                >
                  Change Password
                </button>
              </div>
              <br />
              <div
                className="div-profile"
              >
                <h5>
                  Add Category
                </h5>
                <div
                  className="form-group profile-form"
                >
                  <input
                    className="form-control"
                    id="category-name"
                    name="categoryName"
                    onChange={[Function]}
                    placeholder="Enter Name"
                    type="text"
                    value=""
                  />
                </div>
                <div>
                  <button
                    className="btn btn-success"
                    onClick={[Function]}
                    type="button"
                  >
                    Add
                  </button>
                </div>
                <br />
                <div
                  className="profile-category-button"
                >
                  <button
                    className="btn btn-default btn-sm"
                    name="categoryModalName"
                    onClick={[Function]}
                    type="button"
                    value={1}
                  >
                    Local Dish
                  </button>
                </div>
              </div>
              <div />
            </main>
          </section>
          <section
            className="col-md-9 profile-tabs"
          >
            <div
              className="div-section"
            >
              <div>
                <ul
                  className="nav nav-tabs nav-fill"
                >
                  <li
                    className="nav-item"
                  >
                    <a
                      aria-current={false}
                      className="nav-link"
                      href="#/dashboard/top-recipes"
                      onClick={[Function]}
                      style={undefined}
                    >
                      Top Recipes
                    </a>
                  </li>
                  <li
                    className="nav-item"
                  >
                    <a
                      aria-current={false}
                      className="nav-link"
                      href="#/dashboard/favorites"
                      onClick={[Function]}
                      style={undefined}
                    >
                      Favorites
                    </a>
                  </li>
                  <li
                    className="nav-item dropdown"
                  >
                    <a
                      aria-expanded="false"
                      aria-haspopup="true"
                      className=" nav-link dropdown-toggle"
                      data-toggle="dropdown"
                      role="button"
                    >
                      My Recipes
                    </a>
                    <div
                      className="dropdown-menu"
                    >
                      <a
                        aria-current={false}
                        className="dropdown-item"
                        href="#/dashboard/my-recipes"
                        onClick={[Function]}
                        style={undefined}
                      >
                        Personal Recipes
                      </a>
                      <a
                        aria-current={false}
                        className="dropdown-item"
                        href="#/dashboard/manage-recipes"
                        onClick={[Function]}
                        style={undefined}
                      >
                        Manage Recipes
                      </a>
                    </div>
                  </li>
                  <li
                    className="nav-item"
                  >
                    <a
                      aria-current={false}
                      className="nav-link"
                      href="#/dashboard/add-recipe"
                      onClick={[Function]}
                      style={undefined}
                    >
                      Add Recipe
                    </a>
                  </li>
                </ul>
              </div>
              <br />
              <div />
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</div>
`;
